### ISSUE TITLE
Proposal: support manual control of `service update` process / canary

### ISSUE THREAD

**Issue by pnickolov** [docker/docker#26160]
When deploying a new version of a container with multiple replicas to live production apps, especially apps under load, operators frequently need to control the process of replacing the old version containers with the new version container - e.g., to check the error logs, end-to-end app performance, etc., before proceeding with the next replica (or next set of replicas).

While it is possible to simply set a long delay using `--update-delay`, this is not enough nor convenient if a number of checks are needed. Supporting pause/unpause of the update process (see #26159) can help in ad-hoc cases, but it is also inconvenient if these checks are part of the formal deployment process.

Proposing to add API and CLI command to manually control advancing the update process, e.g., `service update-step <service> add=<n_new> remove=<n_old>` which allows the operator to manually control the service update process but specifying how many new replicas are to be created and how many of the old ones to be removed. This may also require adding `--update-mode=manual` , `--start-paused` option to `service update`, or a special value to `--update-delay` in order to tell the daemon that the updates will be controlled manually.

Please comment on the need and approach/alternatives. If this is a desirable change, I would be happy to contribute

**Comment by aaronlehmann** [comment]
I like the idea of being able to manually step a paused update.

I think that rather than specifying separate numbers of tasks to add and remove, it would be better to specify one number with how many "steps" to take. The updater works by atomically stopping an old task and starting a new one at the same time. There is work underway to make this more flexible and allow the new task to start before the old one is stopped: https://github.com/docker/swarmkit/pull/1383. Telling the updater to just do `n` steps would fit much better with the current model, and could leverage this enhancement. Giving the user fine-grained control over how many tasks to start and stop could put the service in a state where it's out of sync with its spec, and certain things like a leader failover that trigger reconciliation would cause some tasks to be killed or created.

**Comment by pnickolov** [comment]
@aaronlehmann, I agree - doing it the way you propose makes sense. It is also simpler as interface and still addresses the need. Thanks! If I understand correctly, if I specify making, let say, 3 steps, the updater will do 3 steps using the normal method (including delay, etc.) and then stop/pause. Correct?

It looks like `update-pause`, `update-unpause` and `update-step N`

It will still require, I think, to have a way to start the update in "manual" mode (or paused); this may also be specified as number of steps to take before stopping (so, it may be 0 or 1; infinite if not specified - looks a bit clunky but it depends on the design in place)

Should this issue be moved to docker/swarmkit ?

**Comment by aaronlehmann** [comment]
> Thanks! If I understand correctly, if I specify making, let say, 3 steps, the updater will do 3 steps using the normal method (including delay, etc.) and then stop/pause. Correct?

Yes.

> It looks like update-pause, update-unpause and update-step N
>
> It will still require, I think, to have a way to start the update in "manual" mode (or paused); this may also be specified as number of steps to take before stopping (so, it may be 0 or 1; infinite if not specified - looks a bit clunky but it depends on the design in place)

Perhaps it would make sense to have a `--pause` flag on `service update` instead of an `update-pause` subcommand. This would actually fit the data model well, since I'm thinking the pause flag should be part of the service spec (see https://github.com/docker/docker/issues/26159#issuecomment-243580196).

In this model:

- `docker service update --pause servicename` would pause an ongoing update
- `docker service update <other flags> --pause servicename` would initiate an update that makes whatever changes are specified, but leave it frozen in a paused state
- `docker service update --resume servicename` would clear the flag that causes the update to stay paused, and the update would proceed automatically from that point.

It would avoid the need for new top-level commands.

WDYT?

> Should this issue be moved to docker/swarmkit ?

👍, these discussions are more relevant to swarmkit.

**Comment by pnickolov** [comment]
@aaronlehmann , thanks. I'll move it to swarmkit.

On having `--pause` and `--resume` as options to `service update`: OK, I understand why this may be preferred. I like that I can pass other `update` flags as part of pause/resume (e.g., change the delay or the rollout mode per docker/swarmkit#1383). 

If I understand correctly, this sub-command approach affects only the CLI; in the API, we should probably keep them as separate operations, though, right?

I also think I understand what you proposed as design, will look through the code and post for discussion. Looks like:

1. having a `paused` flag (which may be set before the update begins or after)

2. having a counter like `n_auto_steps`, which is infinity in the default case (when in fully automatic mode) or can be set to a finite value by the `update --steps=3` command

3. each container replacement completed decrements the `n_auto_steps`; if it reaches 0, the `paused` flag is set (which will use a pause mechanism that already exists internally)

4. the `paused` flag and `n_auto_steps` fields are kept outside of the specification, so that changing them will not trigger new update

Is this in the ballpark?

**Comment by aaronlehmann** [comment]
> If I understand correctly, this sub-command approach affects only the CLI; in the API, we should probably keep them as separate operations, though, right?

I think they should also be treated as an update at the API level; see below.

> I also think I understand what you proposed as design, will look through the code and post for discussion. Looks like:

> 1. having a paused flag (which may be set before the update begins or after)

I'm envisioning two pause flags. One is the user's control point over whether to put the update into a paused state. As a flag under the user's control, this would go inside the service spec. There is also a field indicating the actual state of the update - this already exists as `Service.UpdateStatus.State` (one of the values is `PAUSED`). This is outside the service spec because it can get set without user intervention, for example when an update encounters a failure.

So having the paused flag set in the spec would cause `Service.UpdateStatus.State` to change to `PAUSED`, and the updater looks to this flag to check if it should continue updating tasks.

> 2. having a counter like n_auto_steps, which is infinity in the default case (when in fully automatic mode) or can be set to a finite value by the update --steps=3 command

> 3. each container replacement completed decrements the n_auto_steps; if it reaches 0, the paused flag is set (which will use a pause mechanism that already exists internally)

To keep things simple, I might do this by putting `steps` in the service spec, and having the updater make `steps` changes and then set `Service.UpdateStatus.State` to `PAUSED`.

> 4. the paused flag and n_auto_steps fields are kept outside of the specification, so that changing them will not trigger new update

The user's preferences about pausing an update and manual stepping do belong in the specification.

On avoiding starting a new update: there are two pieces to this. One is the updater itself. Right now it only considers changes to `Spec.Task` and `Spec.Endpoint`, and won't interrupt the current update unless one of those changed. Arguably, this is a bug, because it means changing something like "update parallelism" during an ongoing update will have no effect.

```go
        if update, ok := u.updates[id]; ok {
                if !update.isServiceDirty(service) {
                        // There's already an update working towards this goal.
                        return
                }
                update.Cancel()
        }
```

Anyway, we would need to change the updater to also consider things like "paused" and "steps" inside the spec when deciding whether to ignore a change to a service or not. Simply replacing the `isServiceDirty` check with a comparison of the whole spec should do the trick.

The other piece is the control API. Right now, any change to the service specification clears `UpdateStatus` on the assumption that this change will trigger a new update. This would need to be refined to avoid doing so when the change isn't actually going to trigger a new update; for example, setting `paused = true` might pause an existing one.

**Comment by yayuntian** [comment]
+1
i hope to see the code merge

**Comment by DoomTaper** [comment]
Is there any update on this ?
